ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,525 @@
+   *   {@link Account}'s id and name as key.
+   * </p>
+   */
+  private class AccountInfoMap {
+    private final Map<String, Account> nameToAccountMap;
+    private final Map<Short, Account> idToAccountMap;
+
+    /**
+     * Constructor.
+     */
+    private AccountInfoMap() {
+      nameToAccountMap = new HashMap<>();
+      idToAccountMap = new HashMap<>();
+    }
+
+    /**
+     * <p>
+     *   Constructs an {@code AccountInfoMap} from a group of {@link Account}s. The {@link Account}s exists
+     *   in the form of a string-to-string map, where the key is the string form of an {@link Account}'s id,
+     *   and the value is the string form of the {@link Account}'s JSON string.
+     * </p>
+     * <p>
+     *   The source {@link Account}s in the {@code accountMap} may have duplicate account Ids or names. The
+     *   constructor makes the best effort to resolve the conflict. Based on the order when iterating on the
+     *   {@link Account}s, the constructor will: 1) override the previously-added {@link Account} if a latter
+     *   one has the same id; and 2) remove the previously-added {@link Account} if a latter one has the same
+     *   name. Once an {@code AccountInfoMap} is constructed, it is guaranteed there is no duplicate id or name.
+     * </p>
+     * @param accountMap A map of {@link Account}s in the form of (accountIdString, accountJSONString).
+     */
+    private AccountInfoMap(Map<String, String> accountMap) {
+      nameToAccountMap = new HashMap<>();
+      idToAccountMap = new HashMap<>();
+      String errorMessage = null;
+      for (Map.Entry<String, String> entry : accountMap.entrySet()) {
+        String idKey = entry.getKey();
+        String valueString = entry.getValue();
+        Account account;
+        try {
+          JSONObject accountJson = new JSONObject(valueString);
+          if (idKey == null) {
+            errorMessage = "Invalid account record when reading accountMap in ZNRecord because idKey=null";
+            accountServiceMetrics.invalidRecordErrorCount.inc();
+            accountServiceMetrics.nullIdKeyErrorCount.inc();
+            break;
+          }
+          account = Account.fromJson(accountJson);
+          if (account.getId() != Short.valueOf(idKey)) {
+            errorMessage =
+                "Invalid account record when reading accountMap in ZNRecord because idKey and accountId do not match. idKey="
+                    + idKey + " accountId=" + account.getId();
+            accountServiceMetrics.invalidRecordErrorCount.inc();
+            accountServiceMetrics.unMatchedIdKeyAndAccountErrorCount.inc();
+            break;
+          }
+        } catch (Exception e) {
+          String message = "Invalid account record when reading accountMap in ZNRecord. key=" + idKey + " valueString="
+              + valueString;
+          accountServiceMetrics.invalidRecordErrorCount.inc();
+          throw new IllegalStateException(message, e);
```

Comment:
I agree with your suggestion. The only question I have is, if we just throw in private methods, it may lose the context for debugging. Surely we will able to recover the call stack, but how about the actual cause to the exception? Back to this example, the reason for catching and rethrowing is exactly for the purpose of recording more information in the exception message.